assembly/src/ast/mod.rs

```diff
@@ -403,12 +430,29 @@ impl ModuleAst {
     }
 }
 
-impl Serializable for ModuleAst {
+/// Implementation of Serializable for ModuleAst.
+/// Serialization of ModuleAsts requires AstSerializationOptions, so
+/// we use a specialized type for serialization.
+pub struct SerializableModuleAst<'a> {
+    module: &'a ModuleAst,
+    options: AstSerializationOptions,
+}
```

Comment:
I'm not 100% sure we need this struct. The way I see it, we have 2 scenarios:

1. Serializing/deserializing individual modules via `to_bytes()` and `from_bytes()` functions. In this case, we can serialize options together with the AST (or have the options be parameters to both - similar to what I mentioned in my previous comment).
2. Serialization/deserialization of modules in libraries. For this, it seems to me that there should be one set of options per library (I can't think of a good use case for why different modules in a library would have different serialization options).

With imports specifically, we already have the info on whether there are any imports to be deserialized (via the number of imports value) - and when we need to know whether imports were in fact serialized, this could come from higher-level components (e.g., library).

Reply:
The reason introduced this extra struct was to retain `ModuleAst` 's implementation of `Serializable`. If we add `SerializationOptions` as a parameter to `write_into`, then the function signature will be wrong.

We can change the serialization function to `to_bytes()` and `from_bytes()` like we do for `ProgramAst`? I did wonder why that hadn't been done already, but I figured there was a good reason for it, so I didn't want to change it.

Reply:
I just tried to play around with this, and it has a bigger blast radius than I thought. `ModuleAst::write_into` is called from `MaslLibrary::write_into` and `Module::write_into` (see assembly/src/library/{mod.rs, masl.rs}`, so we have the following options:

- Give up `Serializable` for `MaslLibrary` and `Module`, or
- Use default serialization options for `ModuleAst::write_into`.

If we go with the default options I assume the default will be `serialize_imports = false`, since that's what we've been doing everywhere so far. However, it does feel counter-intuitive that the default option is to not serialize the entire datastructure, so someone is bound to try to access the imports of a deserialized module, and then spending time figuring out why they have suddenly disappeared.

Reply:
What about doing something like this:

We keep `Serializable`/`Deserializable` implemented for `MaslLibrary` and always serialize/deserialize options together with the library.

We give up `Serializable`/`Deserializable` implementations for `ModuleAst` and instead manually implement `write_into()` and `read_from()` on `ModuleAst` struct. The signatures for these methods would be as follows:

```Rust
pub fn write_into<W: ByteWriter>(&self, target: &mut W, opts: AstSerdeOptions);
pub fn read_from<R: ByteReader>(source: &mut R, opts: AstSerdeOptions) -> Result<Self, DeserializationError>
```
Then, we can call these methods from MASL library serializer/deserializer and pass the appropriate options, and also use them from `to_bytes()` and `from_bytes()`. In the latter case, `to_bytes()` and `from_bytes()` would be responsible for serializing/deserializing the options themselves.

Would this work, or would it result in some complications?

Reply:
I'm fine with that suggestion, except for this point:

> In the latter case, `to_bytes()` and `from_bytes()` would be responsible for serializing/deserializing the options themselves.

Is there any point to having `ModuleAst::to_bytes()` and `ModuleAst::from_bytes()` if `ModuleAst` doesn't implement `Serializable` and `Deserializable`? 

If so, the signatures would have to include an `AstSerdeOptions` parameter, or use a default option.

Reply:
> If so, the signatures would have to include an `AstSerdeOptions` parameter, or use a default option.

Yes, `to_bytes()` would need this options, but `from_bytes()` would deserialize options from the byte stream.

> Is there any point to having `ModuleAst::to_bytes()` and `ModuleAst::from_bytes()` if `ModuleAst` doesn't implement `Serializable` and `Deserializable`?

Yes, we would want to be able to serialize individual modules (probably to something like `.masb` files) in the future. But that doesn't need to be a part of this PR.

Reply:
OK, got it. New commit coming up shortly.

Reply:
Fixed.

I don't suppose we want `write_into()` and `read_from()` functions for `ProgramAst` too?

Reply:
Great!

> I don't suppose we want `write_into()` and `read_from()` functions for `ProgramAst` too?

No, I don't think we need them.

I've just merged the re-exports PR - so, now there are conflicts (sorry!). Could you please resolve them?

Reply:
No problem. I'd like to add a couple of tests to this PR as well. I'll ping you when it's ready to be merged.